### PR TITLE
Use extension methods in place of method properties

### LINKED ISSUE #104: Use interface methods without having to implement properties
We now use extension methods to provide the syntactic sugar for calling methods on an interface. For example, ICustomRay required you to implement `showRay` and `hideRay` as properties. These are now provided automagically as `this.ShowRay` and `this.HideRay`.

### PR DESCRIPTION
**Pull request by amirebrahimi**
Resolves #104 

This is the big refactor that decentralizes ConnectInterfaces/DisconnectInterfaces. It also cleans up field usage and module instantiation.

IInterfaceConnector/ILateBindInterfaceMethods is where the a good bit of the magic happens.

### CODE REVIEW

File: Scripts/Core/Interfaces/IInterfaceConnector.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿using UnityEngine;
+
+namespace UnityEditor.Experimental.EditorVR.Core
+{
+	public interface IInterfaceConnector
```
Comment (mtschoen-unity) [inline Scripts/Core/Interfaces/IInterfaceConnector.cs]: Should this be internal? Otherwise, it needs documentation.

File: Scripts/Interfaces/IInstantiateUI.cs [hunk 2]
```diff
@@ -4,24 +4,33 @@
...
+
+	public static class IInstantiateUIMethods
+	{
+		internal delegate GameObject InstantiateUIDelegate(GameObject prefab, Transform parent = null, bool worldPositionStays = true);
+
+		internal static InstantiateUIDelegate instantiateUI { get; set; }
+
 		/// <summary>
-		/// Method provided by the system for instantiating UI
+		/// Method for instantiating UI
```
Comment (mtschoen-unity) [inline Scripts/Interfaces/IInstantiateUI.cs]: Maybe we want to keep "provided by the system?"

File: Scripts/Interfaces/IMoveCameraRig.cs [hunk 3]
```diff
@@ -4,21 +4,28 @@
...
+
+	public static class IMoveCameraRigMethods
+	{
+		internal delegate void MoveCameraRigDelegate(Vector3 position, Vector3? viewDirection = null);
+
+		internal static MoveCameraRigDelegate moveCameraRig { get; set; }
+
 		/// <summary>
-		/// Move the camera rig using the standard interpolation provided by the system
+		/// Method signature for moving the camera rig
```
Comment (mtschoen-unity) [inline Scripts/Interfaces/IMoveCameraRig.cs]: Remove "signature"

File: Scripts/Interfaces/IPlaceSceneObject.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+	public interface IPlaceSceneObject
+	{
+	}
+
+	public static class IPlaceSceneObjectMethods
+	{
+		internal static Action<Transform, Vector3> placeSceneObject { get; set; }
+
+		/// <summary>
+		/// Delegate used to place objects in the scene/MiniWorld
```
Comment (mtschoen-unity) [inline Scripts/Interfaces/IPlaceSceneObject.cs]: Replace "Delegate" with "Method"

File: Scripts/Interfaces/ISelectTool.cs [hunk 5]
```diff
@@ -9,13 +9,25 @@ namespace UnityEditor.Experimental.EditorVR
...
 	public interface ISelectTool
 	{
+	}
+
+	public static class ISelectToolMethods
+	{
+		internal static Func<Transform, Type, bool> selectTool { get; set; }
+
 		/// <summary>
 		/// Delegate used to select tools from the menu
```
Comment (mtschoen-unity) [inline Scripts/Interfaces/ISelectTool.cs]: Replace "Delegate" with "Method"

File: Scripts/Interfaces/ISetHighlight.cs [hunk 6]
```diff
@@ -4,23 +4,30 @@
...
+
+	public static class ISetHighlightMethods
+	{
+		internal delegate void SetHighlightDelegate(GameObject go, bool active, Transform rayOrigin = null, Material material = null);
+
+		internal static SetHighlightDelegate setHighlight { get; set; }
+
 		/// <summary>
-		/// Method provided by the system to add or remove a highlight on an object
+		/// Method signature highlighting objects
```
Comment (mtschoen-unity) [inline Scripts/Interfaces/ISetHighlight.cs]: Remove "Signature"

File: Scripts/Interfaces/IUsesRaycastResults.cs [hunk 7]
```diff
@@ -9,12 +9,24 @@ namespace UnityEditor.Experimental.EditorVR
...
 	public interface IUsesRaycastResults
 	{
+	}
+
+	public static class IUsesRaycastResultsMethods
+	{
+		internal static Func<Transform, GameObject> getFirstGameObject { get; set; }
+
 		/// <summary>
 		/// Delegate used to test hover/intersection
```
Comment (mtschoen-unity) [inline Scripts/Interfaces/IUsesRaycastResults.cs]: Replace "Delegate" with "Method"

File: Scripts/Utilities/ObjectUtils.cs [hunk 8]
```diff
@@ -61,7 +62,7 @@ public static bool IsInLayer(GameObject o, string s)
 		public static GameObject CreateEmptyGameObject(String name = null, Transform parent = null)
 		{
 			GameObject empty = null;
-			if (string.IsNullOrEmpty(name))
+			if (String.IsNullOrEmpty(name))
```
Comment (mtschoen-unity) [inline Scripts/Utilities/ObjectUtils.cs]: Why this change? Also why String above and not string?

Come to think of it, this method is a little odd to me. Outside of EVR it just returns null?

File: Scripts/Utilities/ObjectUtils.cs [hunk 9]
```diff
@@ -224,6 +225,13 @@ public static IEnumerable<Type> GetExtensionsOfClass(Type type)
 			return Enumerable.Empty<Type>();
 		}
 
+		public static bool IsSameAssembly<T>(object obj)
```
Comment (mtschoen-unity) [inline Scripts/Utilities/ObjectUtils.cs]: This isn't used. Also, I didn't think it was new...

File: Scripts/Interfaces/ISetManipulatorsVisible.cs [hunk 10]
```diff
@@ -8,10 +8,20 @@ namespace UnityEditor.Experimental.EditorVR
...
+
+	public static class ISetManipulatorsVisibleMethods
+	{
+		internal static Action<ISetManipulatorsVisible, bool> setManipulatorsVisible { get; set; }
+
 		/// <summary>
 		/// Show or hide the manipulator(s)
 		/// </summary>
-		Action<ISetManipulatorsVisible, bool> setManipulatorsVisible { set; }
+		public static void SetManipulatorsVisible(this ISetManipulatorsVisible obj, ISetManipulatorsVisible requester, bool visibility)
```
Comment (dunity) [inline Scripts/Interfaces/ISetManipulatorsVisible.cs]: missing summary parameters

File: Scripts/Interfaces/ISetTooltipVisibility.cs [hunk 11]
```diff
@@ -8,15 +8,31 @@ namespace UnityEditor.Experimental.EditorVR
...
+	public static class ISetTooltipVisibilityMethods
+	{
+		internal static Action<ITooltip> showTooltip { get; set; }
+		internal static Action<ITooltip> hideTooltip { get; set; }
+
 		/// <summary>
 		/// Show the given Tooltip
 		/// </summary>
-		Action<ITooltip> showTooltip { set; }
+		public static void ShowTooltip(this ISetTooltipVisibility obj, ITooltip tooltip)
```
Comment (dunity) [inline Scripts/Interfaces/ISetTooltipVisibility.cs]: missing summary parameter

File: Scripts/Interfaces/ISetTooltipVisibility.cs [hunk 12]
```diff
@@ -8,15 +8,31 @@ namespace UnityEditor.Experimental.EditorVR
...
+		{
+			if (showTooltip != null)
+				showTooltip(tooltip);
+		}
 
 		/// <summary>
 		/// Hide the given Tooltip
 		/// </summary>
-		Action<ITooltip> hideTooltip { set; }
+		public static void HideTooltip(this ISetTooltipVisibility obj, ITooltip tooltip)
```
Comment (dunity) [inline Scripts/Interfaces/ISetTooltipVisibility.cs]: missing summary parameter

File: Scripts/Interfaces/IUsesGameObjectLocking.cs [hunk 13]
```diff
@@ -9,15 +9,32 @@ namespace UnityEditor.Experimental.EditorVR
...
+	public static class IUsesGameObjectLockingMethods
+	{
+		internal static Action<GameObject, bool> setLocked { get; set; }
+		internal static Func<GameObject, bool> isLocked { get; set; }
+
 		/// <summary>
 		/// Set a GameObject's locked status
 		/// </summary>
-		Action<GameObject, bool> setLocked { set; }
+		public static void SetLocked(this IUsesGameObjectLocking obj, GameObject go, bool locked)
```
Comment (dunity) [inline Scripts/Interfaces/IUsesGameObjectLocking.cs]: missing summary parameter

File: Scripts/Interfaces/IUsesGameObjectLocking.cs [hunk 14]
```diff
@@ -9,15 +9,32 @@ namespace UnityEditor.Experimental.EditorVR
...
+		{
+			if (setLocked != null)
+				setLocked(go, locked);
+		}
 
 		/// <summary>
 		/// Check whether a GameObject is locked
 		/// </summary>
-		Func<GameObject,bool> isLocked { set; }
+		public static bool IsLocked(this IUsesGameObjectLocking obj, GameObject go)
```
Comment (dunity) [inline Scripts/Interfaces/IUsesGameObjectLocking.cs]: missing summary parameter

File: Scripts/Interfaces/IUsesSpatialHash.cs [hunk 15]
```diff
@@ -9,15 +9,30 @@ namespace UnityEditor.Experimental.EditorVR
...
+	public static class IUsesSpatialHashMethods
+	{
+		internal static Action<GameObject> addToSpatialHash { get; set; }
+		internal static Action<GameObject> removeFromSpatialHash { get; set; }
+
 		/// <summary>
 		/// Add all renderers of a GameObject (and its children) to the spatial hash for queries, direct selection, etc.
 		/// </summary>
-		Action<GameObject> addToSpatialHash { set; }
+		public static void AddToSpatialHash(this IUsesSpatialHash obj, GameObject go)
```
Comment (dunity) [inline Scripts/Interfaces/IUsesSpatialHash.cs]: missing summary parameter

File: Scripts/Interfaces/IUsesSpatialHash.cs [hunk 16]
```diff
@@ -9,15 +9,30 @@ namespace UnityEditor.Experimental.EditorVR
...
+		{
+			if (addToSpatialHash != null)
+				addToSpatialHash(go);
+		}
 
 		/// <summary>
 		/// Remove all renderers of a GameObject (and its children) from the spatial hash
 		/// </summary>
-		Action<GameObject> removeFromSpatialHash { set; }
+		public static void RemoveFromSpatialHash(this IUsesSpatialHash obj, GameObject go)
```
Comment (dunity) [inline Scripts/Interfaces/IUsesSpatialHash.cs]: missing summary parameter

File: Scripts/Interfaces/IUsesViewerBody.cs [hunk 17]
```diff
@@ -9,10 +9,22 @@ namespace UnityEditor.Experimental.EditorVR
...
+
+	public static class IUsesViewerBodyMethods
+	{
+		internal static Func<Transform, bool> isOverShoulder { get; set; }
+
 		/// <summary>
 		/// Returns whether the specified transform is over the viewer's shoulders and behind the head
 		/// </summary>
-		Func<Transform, bool> isOverShoulder { set; }
+		public static bool IsOverShoulder(this IUsesViewerBody obj, Transform transform)
```
Comment (dunity) [inline Scripts/Interfaces/IUsesViewerBody.cs]: missing summary parameter

File: Scripts/Input/SixenseInputToEvents.cs [hunk 18]
```diff
@@ -1,6 +1,7 @@
 ﻿#if UNITY_EDITOR
 using System;
 using UnityEditor.Experimental.EditorVR;
+using UnityEditor.Experimental.EditorVR.Core;
```
Comment (amirebrahimi) [inline Scripts/Input/SixenseInputToEvents.cs]: Showing as used for me. VRView is there.

Comment (mtschoen-unity) [inline Scripts/Input/SixenseInputToEvents.cs]: My mistake. I forgot the rest of the code is commented out for me

File: Scripts/Core/Interfaces/ILateBindInterfaceMethods.cs [hunk 19]
```diff
@@ -0,0 +1,9 @@
+﻿using UnityEngine;
+
+namespace UnityEditor.Experimental.EditorVR.Core
+{
+	public interface ILateBindInterfaceMethods<T> where T : class
```
Comment (amirebrahimi) [inline Scripts/Core/Interfaces/ILateBindInterfaceMethods.cs]: Made it internal, so no summary needed.

File: Scripts/Interfaces/IInstantiateMenuUI.cs [hunk 20]
```diff
@@ -9,13 +9,24 @@ namespace UnityEditor.Experimental.EditorVR
...
+
 		/// <summary>
 		/// Instantiate custom menu UI on a proxy
-		/// Transform = Ray origin
-		/// IMenu = Prefab (with IMenu component) to instantiate
-		/// Returns an instantiated UI GameObject
 		/// </summary>
-		Func<Transform, IMenu, GameObject> instantiateMenuUI { set; }
+		/// <param name="rayOrigin">The ray origin of the proxy that this menu is being instantiated from</param>
+		/// <param name="menuPrefab">The prefab (with an IMenu component) to instantiate</param>
```
Comment (amirebrahimi) [inline Scripts/Interfaces/IInstantiateMenuUI.cs]: Not sure what you mean on this one. The param names match the parameters and are in the correct order.

File: Scripts/Core/EditorVR.cs [hunk 21]
```diff
@@ -335,11 +326,83 @@ void ProcessInput(HashSet<IProcessInput> processedInputs, ConsumeControlDelegate
...
 		{
-			T module = ObjectUtils.AddComponent<T>(gameObject);
-			m_Interfaces.ConnectInterfaces(module);
-			return module;
+			Nested nested = null;
+			if (!m_NestedModules.TryGetValue(type, out nested))
+			{
+				nested = (Nested)Activator.CreateInstance(type);
+				m_NestedModules.Add(type, nested);
+
```
Comment (amirebrahimi) [inline Scripts/Core/EditorVR.cs]: This isn't showing as addressed because of the line number, but has been.

**Comment by amirebrahimi** [comment]
@dunity / @mtschoen-unity - I've addressed your PR comments

File: Scripts/Interfaces/IInstantiateMenuUI.cs (hunk 20 above)
Comment (dunity) [inline Scripts/Interfaces/IInstantiateMenuUI.cs]: You're correct; I must have meant this for another function.

File: Scripts/Core/InterfaceConnectors/HierarchyModuleConnector.cs [hunk 22]
```diff
@@ -23,12 +22,10 @@ public void ConnectInterface(object obj, Transform rayOrigin = null)
 			}
 
 			public void DisconnectInterface(object obj)
-			{
-				var evrHierarchyModule = evr.GetModule<HierarchyModule>();
-
-				var usesHierarchy = obj as IUsesHierarchyData;
+			{				var usesHierarchy = obj as IUsesHierarchyData;
```
Comment (dunity) [inline Scripts/Core/InterfaceConnectors/HierarchyModuleConnector.cs]: move to line below

File: Scripts/Core/InterfaceConnectors/ActionsModuleConnector.cs [hunk 23]
```diff
@@ -9,40 +9,44 @@ class ActionsModuleConnector : Nested, IInterfaceConnector
...
-				var evrMenus = evr.GetNestedModule<Menus>();
-
-				var toolActions = obj as IActions;
-				if (toolActions != null)
+				var actionsModule = evr.GetModule<ActionsModule>();
+				if (actionsModule)
 				{
-					var actions = toolActions.actions;
-					foreach (var action in actions)
+					var menuActions = actionsModule.menuActions;
```
Comment (dunity) [inline Scripts/Core/InterfaceConnectors/ActionsModuleConnector.cs]: menuActions & evrMenus should be cached inside the toolActions null check block

**Comment by amirebrahimi** [comment]
@dunity - I didn't see any new changes you were requesting. Can you point me to those?
